### PR TITLE
[BEAM-13156][Playground] Implement streaming output for the run code logs

### PR DESCRIPTION
**Pull request by AydarZaynutdinov**
implementation of streaming output for run code output;

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/backend/cmd/server/controller.go [hunk 1]
```diff
@@ -87,19 +87,48 @@ func (controller *playgroundController) CheckStatus(ctx context.Context, info *p
...
 
-//GetRunOutput is returning output of execution for specific pipeline by PipelineUuid
+// GetRunOutput is returning output of execution for specific pipeline by PipelineUuid
 func (controller *playgroundController) GetRunOutput(ctx context.Context, info *pb.GetRunOutputRequest) (*pb.GetRunOutputResponse, error) {
 	pipelineId := info.PipelineUuid
+	runOutputLastIndex, err := controller.cacheService.GetValue(ctx, uuid.MustParse(pipelineId), cache.RunOutputIndex)
+	if err != nil {
+		logger.Errorf("%s: GetRunOutput(): cache.GetValue: error: %s", pipelineId, err.Error())
+		return nil, errors.NotFoundError("GetRunOutput", "there is no run output for pipelineId: "+pipelineId+", subKey: cache.RunOutputLastIndex")
+	}
```
Comment (KhaninArtur) [inline playground/backend/cmd/server/controller.go:97]: What happens if it is the first time we try to retrieve the `runOutputLastIndex`? Does it become equal to 0?

**Review by KhaninArtur — CHANGES_REQUESTED** [review]

File: playground/backend/cmd/server/controller.go [hunk 2]
```diff
@@ -87,19 +87,48 @@ func (controller *playgroundController) CheckStatus(ctx context.Context, info *p
...
 }
 
-//GetRunOutput is returning output of execution for specific pipeline by PipelineUuid
+// GetRunOutput is returning output of execution for specific pipeline by PipelineUuid
 func (controller *playgroundController) GetRunOutput(ctx context.Context, info *pb.GetRunOutputRequest) (*pb.GetRunOutputResponse, error) {
 	pipelineId := info.PipelineUuid
+	runOutputLastIndex, err := controller.cacheService.GetValue(ctx, uuid.MustParse(pipelineId), cache.RunOutputIndex)
+	if err != nil {
+		logger.Errorf("%s: GetRunOutput(): cache.GetValue: error: %s", pipelineId, err.Error())
+		return nil, errors.NotFoundError("GetRunOutput", "there is no run output for pipelineId: "+pipelineId+", subKey: cache.RunOutputLastIndex")
```
Comment (ilya-kozyrev) [inline playground/backend/cmd/server/controller.go:96]: why we can't use the standard format way without string concatenation?

**Review by ilya-kozyrev — CHANGES_REQUESTED** [review]
Firstly fix refactoring, and while it is internal, can you switch the target for PR to the refactoring branch?

File: playground/backend/cmd/server/controller.go (hunk 1 above)
Comment (AydarZaynutdinov) [reply playground/backend/cmd/server/controller.go]: If we are trying to receive some value from cache (using key and subKey), but cache doesn't contain the combination of (key:subkey) then we will receive an error (example from `local_cache`):
```
value, found := lc.items[pipelineId][subKey]
	if !found {
		lc.RUnlock()
		return nil, fmt.Errorf("value with pipelineId: %s and subKey: %s not found", pipelineId, subKey)
	}
```
Does it answer your question?

File: playground/backend/cmd/server/controller.go (hunk 2 above)
Comment (AydarZaynutdinov) [reply playground/backend/cmd/server/controller.go]: Is this question still actual?

File: playground/backend/cmd/server/controller.go (hunk 1 above)
Comment (KhaninArtur) [reply playground/backend/cmd/server/controller.go]: I got it – I mean, if we have such a combination of (key:subKey) in cache but we never retrieved it before, what happens with `runOutputLastIndex`?

Comment (AydarZaynutdinov) [reply playground/backend/cmd/server/controller.go]: We put `0` to cache as a value for `{pipelineId}:cache.RunOutputIndex` combination in `RunCode()` method. So if we call `GetRunOutput()` after `RunCode()` then cache should contain value for `{pipelineId}:cache.RunOutputIndex`.

**Review by ilya-kozyrev — CHANGES_REQUESTED** [review]

File: playground/backend/internal/streaming/run_output_writer.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	CacheService cache.Cache
+	PipelineId   uuid.UUID
+}
+
+// Write writes len(p) bytes from p to cache with cache.RunOutput subKey
+func (row *RunOutputWriter) Write(p []byte) (int, error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+	str := string(p)
```
[the remark this answers is not in the document]
Comment (AydarZaynutdinov) [reply playground/backend/internal/streaming/run_output_writer.go]: Yes, we can. Changed.

File: playground/backend/internal/streaming/run_output_writer.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+// RunOutputWriter is used to write the run step's output to cache as a stream.
+type RunOutputWriter struct {
+	Ctx          context.Context
+	CacheService cache.Cache
+	PipelineId   uuid.UUID
+}
+
+// Write writes len(p) bytes from p to cache with cache.RunOutput subKey
+func (row *RunOutputWriter) Write(p []byte) (int, error) {
```
[the remark this answers is not in the document]
Comment (AydarZaynutdinov) [reply playground/backend/internal/streaming/run_output_writer.go:41]: Added comments for this method.

File: playground/backend/internal/code_processing/code_processing.go [hunk 5]
```diff
@@ -55,32 +57,42 @@ func Process(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycl
...
 		return
 	}
 
 	// compile
 	logger.Infof("%s: Compile() ...\n", pipelineId)
-	compileCmd := exec.Compile(ctxWithTimeout)
-	go processCmd(compileCmd, successChannel, errorChannel, dataChannel)
+	compileCmd := executor.Compile(ctxWithTimeout)
+	go func(cmd *exec.Cmd, successCh chan bool, errCh chan error, dataCh chan interface{}) {
+		// TODO separate stderr from stdout
```
Comment (daria-malkova) [inline playground/backend/internal/code_processing/code_processing.go:84]: It is not the best idea to leave to-do's, most likely some reviewer ask to to do it

File: playground/backend/internal/code_processing/code_processing.go [hunk 6]
```diff
@@ -92,14 +104,27 @@ func Process(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycl
...
 
 	// run
 	logger.Infof("%s: Run() ...\n", pipelineId)
-	runCmd := exec.Run(ctxWithTimeout)
-	go processCmd(runCmd, successChannel, errorChannel, dataChannel)
+	runCmd := executor.Run(ctxWithTimeout)
+
+	var runError bytes.Buffer
+	runOutput := &streaming.RunOutputWriter{Ctx: ctxWithTimeout, CacheService: cacheService, PipelineId: pipelineId}
+	runCmd.Stderr = &runError
```
Comment (daria-malkova) [inline playground/backend/internal/code_processing/code_processing.go:115]: why we take this this by reference and the one below by value?

File: playground/backend/internal/streaming/run_output_writer.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+// RunOutputWriter is used to write the run step's output to cache as a stream.
+type RunOutputWriter struct {
+	Ctx          context.Context
+	CacheService cache.Cache
+	PipelineId   uuid.UUID
+}
+
+// Write writes len(p) bytes from p to cache with cache.RunOutput subKey.
```
Comment (daria-malkova) [inline playground/backend/internal/streaming/run_output_writer.go:32]: maybe change writes to sends?

**Review by daria-malkova — CHANGES_REQUESTED** [review]

File: playground/backend/internal/code_processing/code_processing.go (hunk 5 above)
Comment (KhaninArtur) [reply playground/backend/internal/code_processing/code_processing.go:85]: Let's leave a Jira ticket number here with a corresponding task

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 8]
```diff
@@ -444,3 +444,82 @@ func TestGetProcessingStatus(t *testing.T) {
 		})
 	}
 }
+
+func TestGetRunOutputLastIndex(t *testing.T) {
```
Comment (KhaninArtur) [inline playground/backend/internal/code_processing/code_processing_test.go:448]: Should we also use `goleak` here?

File: playground/backend/internal/streaming/run_output_writer_test.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		check   func() bool
+		want    int
+		wantErr bool
+	}{
+		{
+			name: "empty string",
```
Comment (KhaninArtur) [inline playground/backend/internal/streaming/run_output_writer_test.go:51]: Could you please add comments for these tests?

**Review by KhaninArtur — CHANGES_REQUESTED** [review]

**Review by ilya-kozyrev — APPROVED** [review]

**Review by KhaninArtur — APPROVED** [review]

**Pull request closed without merging**